dotnet/src/webdriver/IScript.cs

```diff
@@ -0,0 +1,57 @@
+// <copyright file="IScript.cs" company="WebDriver Committers">
+// Licensed to the Software Freedom Conservancy (SFC) under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The SFC licenses this file
+// to you under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Threading.Tasks;
+using WebDriverBiDi.Log;
+
+namespace OpenQA.Selenium
+{
+    /// <summary>
+    /// Defines an interface providing access to WebDriverBiDi events in script and log domains.
+    /// </summary>
+    [Obsolete("This class is in beta and may change in future releases.")]
+    public interface IScript
+    {
+        /// <summary>
+        /// Add and remove handlers for console messages.
+        /// </summary>
+        [Obsolete("This event is in beta and may change in future releases.")]
+        event EventHandler<EntryAddedEventArgs> ConsoleMessageHandler;
+
+        /// <summary>
+        /// Add and remove handlers for console messages.
+        /// </summary>
+        [Obsolete("This event is in beta and may change in future releases.")]
+        event EventHandler<EntryAddedEventArgs> JavaScriptErrorHandler;
+
+        /// <summary>
+        /// Asynchronously starts monitoring for console and JavaScript log entries.
+        /// </summary>
+        /// <returns>A task object representing the asynchronous operation.</returns>
+        [Obsolete("This task is in beta and may change in future releases.")]
+        Task StartMonitoringLogEntries();
```

Comment:
What's the need for exposing start/stop API?  I would  probably just start when the first event handler is added and stop when the last one is removed.

Reply:
Yeah, it seems less intuitive to me that you need to register and then explicitly call to start monitoring, but apparently that's what is expected in the .NET world 🤷

Reply:
This thread promises to be interesting, and this thread is very important.

> but apparently that's what is expected in the .NET world

No, native events is a standard implementation of pub/sub pattern, or observable pattern, or any other notification based pattern. Events born at the beginning of .net. Technically you are able to know when somebody subscribes/unsubscribes on the event.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/add - pretty easy.

Now, in 2024, let's think about it deeper:
- `add` method is `sync`, but we need to call another `async` method. Sync-over-async problem.
- OK, we know how to call `async` methods from `sync` context even when Microsoft doesn't recommend it.
- What about async event handlers? User wants to call `async` methods in theirs event handlers. This is `async void` problem. Again, Microsoft doesn't recommend it. Ah, this is exceptional case, async event handlers... move on.
- Now, when user is able to define sync/async event handlers, we want to make that all events are dispatched to users, and users handlers are finished. The problem here is how to `await async void`? Kind of impossible. There is tricky trick, but this trick is very rare in .net world. Nobody wants to code workarounds.

How to be?
```csharp
await driver.OnConsoleLogEntryAsync(e => Console.WriteLine(e));

await driver.OnBeforeRequestAsync(async req => await req.ContinueAsync());
```

Why? It is fully manageable!

Reply:
Yeah, my first implementation had private variables for the handlers and then reference them to add/remove

```csharp
 public event EventHandler<EntryAddedEventArgs> ConsoleMessageHandler
        {
            add
            {
                bool startMonitoring = consoleMessageHandler == null;
                consoleMessageHandler += value;
                if (startMonitoring && !isMonitoringStarted)
                {
                    Task.Run(StartMonitoring);
                }
            }
            remove => consoleMessageHandler -= value;
        }
```
But from what I've seen, .NET practices leans toward explicitness and clarity which means not assuming that adding or removing events will start and stop monitoring. (ChatGPT agrees with this 😁)

I'm not following all of your comment, though (and I'm not well-versed in async approach in general). Is there a use case for where we want to make one of these events manage async actions?

Reply:
For user it doesn't matter what you do behind: starting, stopping, laughing. He just subscribe on the event, what means he wants to get notifications every time when the event is occurred.

Use cases for async event handlers?.. My initial example can be considered as use case:
```csharp
await driver.OnBeforeRequestAsync(async req => await req.ContinueAsync());
```

Reply:
I guess my question is whether there is something that can't be done with a sync method, or if it's just that people want to do everything async on principle?

Reply:
With respect to explicit start/stop:

So I took a look at PuppeteerCSharp and Playwright .NET implementations; I don't want to take what they are doing as necessary indications of proper, idiomatic .NET behavior, but, neither of them require an extra step to start monitoring.

I generally see our priorities as:
1. Language idiomatic
2. Matches ease of use in alternative projects
3. Consistent across Selenium bindings

I agree that explicit stop/starts follows the convention of explicitness, but that neither other project requires it and none of the other bindings will do it this way, I'm leaning toward having adding the event start it, but keep the methods for starting and stopping public in case the user wants more control.

Reply:
As far as the async implementation, I did some reading, and yes, we need to support it, I'll see about updating this code.

Reply:
Ok, looks like Async works out of the box...

```
            driver.Script.ConsoleMessageHandler += async (_, msg) =>
            {
                if (msg.Level == WebDriverBiDi.Log.LogLevel.Info)
                {
                    logs.Add(msg);
                }
                await Task.Delay(1000);
            };
```

Reply:
Your example is known `async void` problem. Just try slightly modified handler:
```csharp
await Task.Delay(1000);
if (msg.Level == WebDriverBiDi.Log.LogLevel.Info)
{
  logs.Add(msg);
}

```